### PR TITLE
rewrite bootstrapping stages

### PR DESCRIPTION
**Pull request by mightyiam**

### CODE REVIEW

File: src/building/bootstrapping.md [hunk 1]
```diff
@@ -4,63 +4,101 @@ This subchapter is about the bootstrapping process.
...
+In following stages the compiler is usually the one that was compiled in the previous stage.
+
+### The stages: how each compiler is obtained
+
+#### Stage 0: the pre-compiled compiler
+
+<!--
+What are the associated dynamic libraries?
+Are they the same as "the set of dependencies" above?
+Why are they qualified as dynamic?
```
Comment (jyn514) [inline src/building/bootstrapping.md:38]: https://en.wikipedia.org/wiki/Dynamic_linker

File: src/building/bootstrapping.md [hunk 2]
```diff
@@ -4,63 +4,101 @@ This subchapter is about the bootstrapping process.
...
+
+For this purpose a python script `x.py` is provided at the root of the repository.
+`x.py` downloads a pre-compiled compiler—the stage 0 compiler—and with it compiles from the current source code a compiler—the stage 1 compiler.
+Additionaly, it may use the stage 1 compiler to compile from the current source code another compiler—the stage 2 compiler.
+Below describes this process in some detail, including the reason for a stage 2 compiler and more.
+
+## The stages of bootstrapping
+
+Each stage involves:
+<!-- What is this set of dependencies? -->
```
Comment (jyn514) [inline src/building/bootstrapping.md:22]: https://rustc-dev-guide.rust-lang.org/building/bootstrapping.html#what-is-a-sysroot

File: src/building/bootstrapping.md [hunk 3]
```diff
@@ -4,63 +4,101 @@ This subchapter is about the bootstrapping process.
...
+Why are they qualified as dynamic?
+-->
+A pre-compiled compiler and its associated dyamic libraries are downloaded.
+By default, it is the current beta release.
+This is the stage 0 compiler.
+
+#### Stage 1: from current code, by an earlier compiler
+
+<!--
+What is `rustbuild` and what is its role?
```
Comment (jyn514) [inline src/building/bootstrapping.md:47]: I've been using this to mean "all of bootstrap except for `bootstrap.py`". I don't think that's written down anywhere.

File: src/building/bootstrapping.md [hunk 4]
```diff
@@ -4,63 +4,101 @@ This subchapter is about the bootstrapping process.
...
+Why should it be mentioned here?
+What part does `std` have in compiling the compiler?
```
Comment (jyn514) [inline src/building/bootstrapping.md:49]: https://rustc-dev-guide.rust-lang.org/building/bootstrapping.html#stages-and-std

File: src/building/bootstrapping.md [hunk 5]
```diff
@@ -4,63 +4,101 @@ This subchapter is about the bootstrapping process.
...
+yet, there is no mention of that.
+May be related to the earlier question about dependencies and/or dynamic libraries.
+-->
+By default, the stage 1 libraries are copied into stage 2, because they are expected to be identical.
+
+The stage 1 compiler is used to compile from current code a compiler.
+This is the stage 2 compiler.
+
+The stage 2 compiler is the first that is both from current code and compiled by a compiler that is of current code.
+The compilers <!-- "and libraries" ? --> obtained by `rustup` and other installation methods are all stage 2.
```
Comment (jyn514) [inline src/building/bootstrapping.md:71]: correct

File: src/building/bootstrapping.md [hunk 6]
```diff
@@ -4,63 +4,101 @@ This subchapter is about the bootstrapping process.
...
+
+If compiling a compiler is necessary, most commonly a stage 1 compiler would suffice:
+`x.py build library/std`.
+
+Between the stage 2 and the stage 1 compiler are subtle differences:
+
+
+<!--
+I don't know much about this topic.
+Where are "symbol names used in the compiler source", please?
```
Comment (jyn514) [inline src/building/bootstrapping.md:81]: https://en.wikipedia.org/wiki/Name_mangling

**Review by camelid — COMMENTED** [review]
I'm not sure if a complete rewrite is a good idea. At the very least, it'll probably be a while until this is reviewed since it's such a big change.

**Comment by mightyiam** [comment]
Alright, I feel I made it more readable in general. And I don't feel I missed anything. I feel it's a step forward all things considered. I don't feel anything important has been removed.

> I'm not sure if a complete rewrite is a good idea. At the very least, it'll probably be a while until this is reviewed since it's such a big change.

It seemed to warrant a rewrite, as I was reading it. So that's what I did.

Here's a suggested workflow for reviewing this:

1. Read the current. As you do, make a list of key points that it addresses.
2. Read this version. As you do, check the items that it addresses from the list you previously made.

Each item should be explained at least well enough. That is, if it is relevant at all.

**Comment by JohnTitor** [comment]
[We've discussed this on the Zulip](https://rust-lang.zulipchat.com/#narrow/stream/196385-t-compiler.2Fwg-rustc-dev-guide/topic/review.20rustc-dev-guide.231327), and we're going not to merge this as we don't think this doesn't seem to have much improvement.
If you have any thoughts about this decision, feel free to drop them here.

Thanks for proposing changes anyway!

**Comment by mightyiam** [comment]
> [We've discussed this on the Zulip](https://rust-lang.zulipchat.com/#narrow/stream/196385-t-compiler.2Fwg-rustc-dev-guide/topic/review.20rustc-dev-guide.231327), and we're going not to merge this as we don't think this doesn't seem to have much improvement. If you have any thoughts about this decision, feel free to drop them here.
> 
> Thanks for proposing changes anyway!

It was written carefully. I'm certain that it's a significant improvement.

**Comment by JohnTitor** [comment]
> It was written carefully. I'm certain that it's a significant improvement.

Could you elaborate on what improvements your changes are bringing? I'm not an English native and it's a bit unclear how these changes improve stuff.

**Comment by mightyiam** [comment]
I just updated this branch yet cannot see that update here. Perhaps re-open?

**Comment by JohnTitor** [comment]
I cannot, re-opening would work.

image

**Comment by mightyiam** [comment]
@JohnTitor #1458

**Pull request closed without merging**